Implement SSR adapter abstraction

andresgutgon opened the pull request:
# What?
Now, the only way of running JS code on Elixir is using NodeJS call, but there are other ways of executing JS, for example, using Bun runtime or, in development, rendering the inertia app using a VideJS dev server. These changes allow users to hook into Inertia's SSR system and use their own adapters.

## TODO
- [x] Keep it simple for existing users. It should be totally transparent
- [x] Setup vite plugin for handling HMR on the backend 
- [x] Generate JS/CSS assets based on the ViteJS manifest 
- [x] Being able to use ESM modules for compiled javascript
- [x] Fix credo ✅
- [x] Fix `dyalizer` errors  ✅
- [x] Make all tests pass ✅
- [x] Use these changes in my app 🚀 
- [x] Make using a ViteJS setup in dev and prod easy. Bring ViteJS helpers in [another package](https://github.com/andresgutgon/vitex)

## Adapters
Coming from [this conversation](https://github.com/inertiajs/inertia-phoenix/issues/37#issuecomment-2758031323) and with @valian advise I wanted to give a try to complete integration between inertia-phoenix and ViteJS as js bundler . 
The idea is to use a `ViteJS` adapter during development and use existing `NodeJS` call in production

--- 

# TO BE DISCUSSED
I know you’ve shared your thoughts on this before, @derrickreimer — and I really appreciate your perspective. I hate to bother you with this again, but I’d truly love to bring the Vite experience into working with Inertia + Phoenix.

If I can manage to pull this off in a clean and maintainable way, and if you feel it aligns with the goals of the project, I’d be honored to contribute it. Totally up to you, of course — just wanted to put it out there. 🙏

andresgutgon committed:
Implement SSR adapter system
Now the only way of running JS code on Elixir is using NodeJS call but
there are other ways of executing JS like for example using Bun runtime
or in development rendering the inertia app using VideJS dev server.
This changes allow users to hook into inertia's ssr system and use their
own adapters.

andresgutgon commented on lib/inertia/ssr.ex:
```diff
@@ -21,6 +24,8 @@ defmodule Inertia.SSR do
   - `:path` - (required) the path to the directory where your `ssr.js` file lives.
   - `:module` - (optional) the name of the Node.js module file. Defaults to "#{@default_module}".
+  - `:esm` - (optional) Use ESM for the generated ssr.js file. Defaults to #{@default_esm}.
+  - `:ssr_adapter` - (optional) the name of the Node.js module file. Defaults to "#{@default_ssr_adapter}
```
Fix this description

andresgutgon commented on lib/inertia/ssr/adapters/node_js.ex:
```diff
@@ -0,0 +1,30 @@
+    module = GenServer.call(Config, :module)
+    esm = GenServer.call(Config, :esm)
+
+    case NodeJS.call({module, :render}, [page], name: supervisor_name(), binary: true, esm: esm) do
```
Maybe no need for a case here. Before it was just calling `NodeJS.call`

derrickreimer commented:
Appreciate the effort, @andresgutgon. 

I'm not fundamentally opposed to having multiple adapters for different front-end build systems. If the past is any indication, I don't expect tool churn in the JavaScript world to slow down anytime soon 😅 . So it's reasonable for this library to not be overly opinionated about what people should use when the JS-bundling flavor of the week changes.

I'll need to give it more thought, but it might make sense to put these adapters into separate packages to keep bloat to a minimum.

In the meantime though, I still encourage folks to really consider if you need more than the simple esbuild setup we have baked in!

andresgutgon commented:
> I'm not fundamentally opposed to having multiple adapters for different front-end build systems.

Awesome! 

> I'll need to give it more thought, but it might make sense to put these adapters into separate packages to keep bloat to a minimum.

Yeah that could be a possibility. Let me see if what I'm trying makes sense and we can decide. I really think is not so much bloat but let's the final result.

I'll keep working on this during the weekends and when I have something to show l'll ping you again:)

> I still encourage folks to really consider if you need more than the simple esbuild setup we have baked in!

I guess that depends on the type of dev you're. In my case I 'm the kind that loves these kind of tooling 😂

andresgutgon commented:
Nothing is working yet but I'm loving this
<img width="679" alt="image" src="https://github.com/user-attachments/assets/84e096d8-0801-41bd-8fe5-e85d1fe73605" />

andresgutgon commented on .gitignore:
```diff
@@ -1,5 +1,6 @@
 # The directory Mix will write compiled artifacts to.
 /_build/
+.elixir_ls/
```
When running dyalizer in my machine this appeared. I assume it should be ignored

andresgutgon commented on config/test.exs:
```diff
@@ -8,7 +8,7 @@ config :inertia, MyAppWeb.Endpoint,
 
 # Print only warnings and errors during test
-config :logger, level: :warning
+config :logger, level: :debug
```
⚠️ Revert thsi

andresgutgon commented on test/inertia_test.exs:
```diff
@@ -106,6 +106,8 @@ defmodule InertiaTest do
 
     start_supervised({Inertia.SSR, path: path})
+    {:ok, _pid} = start_supervised({Inertia.SSR, path: path})
+    Process.sleep(100)
```
⚠️ Revert this change

andresgutgon commented:
@derrickreimer I think i made good progress here. Next steps

1. Test the changes on my app
2. Do an app example pointing to these changes but using standard ESBuild setup

But at this point tests are not broken which is good and the internal DX for adapters I think is nice. Let me know

```elixir
defmodule Inertia.SSR do
  alias Inertia.SSR.{Config, Adapter}
  alias Inertia.SSR.Adapters.{Bootstrap, NodeJS}

  @moduledoc """
  Supervisor for SSR support in Inertia views.
  """
  use Supervisor

  def start_link(init_arg) do
    Supervisor.start_link(__MODULE__, init_arg, name: __MODULE__)
  end

  def init(opts) do
    {adapter, config} = Bootstrap.fetch_adapter(opts: opts, default_adapter: NodeJS)
    children = [{Config, adapter: adapter, config: config}] ++ adapter.children(config)
    Supervisor.init(children, strategy: :one_for_one)
  end

  @spec call(Adapter.page()) :: Adapter.ssr_result()
  def call(page) do
    Config.call(page)
  end
end
```

andresgutgon commented:
Ok vitejs plugin is working fine. The nice thing about using Vite dev server for the SSR rendering is that:
1. We only need one vite bundler running in dev
2. We don't need a NodeJS process running in Elixir (only to call Vite dev server)

And integration looks clean
<img width="706" alt="image" src="https://github.com/user-attachments/assets/86b4e020-b8a6-43f0-a0d4-3f7ae585bc0d" />

andresgutgon commented:
Ok, while adding a test for supporting ESM I saw that 
Elixir 15 -> nodejs 18.x
Elixir 16 -> nodejs 20.x

This is LTS table for those versions
| Version | Codename  | Release Date | Active LTS Start | Maintenance Start | End-of-Life (EOL) |
|---------|-----------|--------------|------------------|-------------------|-------------------|
| 18.x    | Hydrogen  | 2022-04-19   | 2022-10-25       | 2023-10-18        | 2025-04-30        |
| 20.x    | Iron      | 2023-04-18   | 2023-10-24       | 2024-10-22        | 2026-04-30        |
| 22.x    | Jod       | 2024-04-24   | 2024-10-29       | 2025-10-21        | 2027-04-30        |
| 23.x    | -         | 2024-10-16   | N/A              | N/A               | 2025-06-01        |

Node 18 is almost 💀 so we could considere moving on. 

Although Elixir 15 is still supported
```markdown

Elixir version | Support
-- | --
1.18 | Bug fixes and security patches
1.17 | Security patches only
1.16 | Security patches only
1.15 | Security patches only
1.14 | Security patches only
```

I tried 1.17.0 and it works. 

The problem I had on CI is that my test fail because node versions prior to 22.x don't auto detect ESM modules. But Node 22.x is the oficial supported version now 😄 

<img width="739" alt="image" src="https://github.com/user-attachments/assets/cb67e00a-6927-4de4-91b8-a2b04eba6120" />


I'll try not to change the CI matrix but wanted to comment on it

andresgutgon commented on README.md:
```diff
@@ -728,6 +732,18 @@ Then, update your config to enable SSR (if you'd like to enable it globally).
+
+### Custom SSR adapter
+
+You can setup your own SSR adapter. For example [vitex](https://github.com/andresgutgon/vitex) is a package that helps with ViteJS development in Phoenix apps. It has a custom SSR adapter for this package (inertia-phoenix) so SSR can be handle in development through Vite Dev server instead of calling a NodeJS process like we do in production.
```
@derrickreimer you commented that maybe would be better if adapters lived outside the Inertia package. And I think I really like this approach because I want to add Vite assets management and that would be too much to put in inertia.

andresgutgon commented on .circleci/config.yml:
```diff
@@ -21,6 +21,8 @@ jobs:
       - checkout
       - run: mix --version
+      - run: node -v
+      - run: npm -v
```
While adding a test for supporting ESM I saw that 
Elixir 15 -> nodejs 18.x
Elixir 16 -> nodejs 20.x

18 is about to lose LTS on 30th of April and 20 next year

andresgutgon commented:
@derrickreimer this is ready for review 🎉

andresgutgon commented:
Super happy with extracting out of inertia the ViteJS adapter. Using `Vitex` users can handle vite dev and production with `NodeJS` adapter 
<img width="768" alt="image" src="https://github.com/user-attachments/assets/1a3ba0e8-7ceb-4e27-9650-e9fe8533c651" />

andresgutgon commented:
@derrickreimer you can review this PR? 🙏

derrickreimer commented:
> @derrickreimer can you review this PR? 🙏

I'm in a pretty busy season at the moment, but I'll do my best to take a look at this soon.

andresgutgon commented:
Okay, I did the last thing I wanted to do. I created an example repo app using my changes but with the standard Inertia setup explained in the README.

https://github.com/andresgutgon/inertia_ssr_adapters


I won't comment more here 🙏. I'll keep moving with my life 😃

andresgutgon commented:
Hey @derrickreimer, hope you’re doing well!

Would you have a chance to review this soon? I’m working off my own fork at the moment, but I noticed you’ve merged some new changes and I’d much rather switch back to using the official package instead of maintaining my own branch.

It’s been a while since I started on this, but I put a lot of thought into the changes. I tried to keep things as non-disruptive as possible for existing users and added tests as well. I’m happy to adjust anything that needs improving—just let me know.

Thanks for taking a look!

andresgutgon commented:
Btw, I noticed Inertia didn't compile unless I add Igniter to my Phoenix `mix.exs` file. Is that expected? 

I see a note in the readme on how to get started with an Inertia project with Igniter, but maybe add a line telling that now Igniter has to be added as a project dependency

andresgutgon commented:
Super quick fix! 👏 

Don't worry, I understand this change can break things. Take your time. It's just that I want to ship this. I think it would be a good addition. Also, I'm quite proud of it. I learned a lot!

andresgutgon commented:
Hi @derrickreimer, it's me again 🙏 I hope you're doing well. Did you have a chance to take a look at these changes?

derrickreimer commented:
> Hi @derrickreimer, it's me again 🙏 I hope you're doing well. Did you have a chance to take a look at these changes?

Apologies, I've not had a chance yet.

andresgutgon commented:
Ok, thanks, Derrick, for the quick response. appreciate

andresgutgon committed:
Merge upstream/main into feature/inertia-vitejs-integration

* upstream/main:
  Fix alias ordering
  Use consistent page_name and previous_page naming
  Fix since tag
  Add support for Scroll props
  Add Once props section to README
  Update CHANGELOG.md
  Add support for Once props
  More helpful prop names for testing behavior
  Fix camelization of keys in deferredProps metadata
  Update CHANGELOG.md
  Update CHANGELOG.md
  run format
  fix wording
  update pages directory in next steps instructions
  add create_pages_directory step

andresgutgon commented:
Hi, it's me again 😅. Happy 2026 @derrickreimer I hope this year is the year of Inertia adapter 💪

vonagam commented:
Just noting that I have used the `intertia` from this PR plus @andresgutgon's `vitex`  (thought the package would need a name change since there is now other `vitex` published) and `phoenix_vite` on back with latest versions of vite/typescript/vue/inertiajs (including it's `@inertiajs/vite` package) on front. Everything worked for me.

Noting that in this PR there is a mention of `ssr_adapter` option for `intertia` config - but it's never read in the code. Right now it's only possible to set adapter by `Inertia.SSR` spec opts.

derrickreimer committed:
Merge remote-tracking branch 'origin/main' into feature/ssr-adapter-system

derrickreimer committed:
Avoid serializing SSR renders through Config GenServer

Previously every Inertia.SSR.call/1 went through a GenServer.call to
Inertia.SSR.Config, which then synchronously invoked the adapter inside
the GenServer process. That serialized all SSR renders behind one
mailbox and defeated the NodeJS worker pool — concurrent requests
queued instead of fanning out.

Store the adapter + config in :persistent_term at supervisor startup
and dispatch from the caller's process. The Inertia.SSR.Config module
is no longer needed and is removed.

derrickreimer committed:
Document :ssr_adapter as a supervisor option, not app config

The previous README example placed `ssr_adapter: MyAdapter` inside the
`config :inertia` block, but that key is read from the `Inertia.SSR`
supervisor's options — never from `Application.get_env(:inertia)`. The
example also lacked a trailing comma, which would have raised a syntax
error if anyone copied it. Remove it from the application config block
and show the correct supervisor-option form in the Custom SSR adapter
section.

Also promote `Inertia.SSR.Adapter` from `@moduledoc false` so the
behaviour callbacks are visible in HexDocs for adapter authors.

derrickreimer committed:
Raise on invalid :ssr_adapter instead of silently falling back

Previously, passing an unloadable module, a non-module value, or a
module that didn't implement the adapter callbacks would silently
fall back to the default NodeJS adapter — masking typos and broken
adapter implementations behind successful boots.

Validate the :ssr_adapter option strictly when supplied and raise an
ArgumentError describing which check failed. A missing :ssr_adapter
still falls back to the default, as before.

derrickreimer committed:
Drop premature Adapters.Config macro

The `Inertia.SSR.Adapters.Config` __using__ macro only existed to delegate
`init/1` to a per-adapter `Config.build/1`, and set a `@name` attribute
that was never read. With only one adapter shipped today, the macro added
indirection without payoff and forced third-party adapter authors to
discover two modules (the behaviour and the macro) instead of one.

Remove the macro and have `Inertia.SSR.Adapters.NodeJS` implement
`@behaviour Inertia.SSR.Adapter` directly with an explicit `init/1`.

derrickreimer committed:
Restore docs on Inertia.SSR public API

The earlier refactor stripped the @doc block off start_link/1 (which
documented :path, :module, and :pool_size with defaults), shrank the
@moduledoc to a single line, and dropped @doc false from init/1 and
call/1 — exposing internal callbacks in the generated docs while
hiding the user-facing options.

Expand the @moduledoc to cover the adapter system, restore the option
docs on start_link/1 (now including :ssr_adapter and :esm), and
re-mark init/1 and call/1 as @doc false.

derrickreimer committed:
Polish: README, Bootstrap signature, NodeJS adapter cleanup

- README: capitalize ECMAScript, add a minimal adapter example to the
  Custom SSR adapter section so authors don't have to reverse-engineer
  the behaviour shape from the source.
- Bootstrap.fetch_adapter: drop the keyword-list destructuring in the
  function head in favour of a conventional 2-arity signature.
- NodeJS adapter:
  - move the @moduledoc to the top of the file, trim the duplicate
    option listing (those now live on Inertia.SSR.start_link/1),
  - drop the unused `require Logger`,
  - replace the misleading `alias Inertia.SSR.Supervisor` (which
    pointed at a non-existent module — it was only ever used as a
    registered process name) with a labeled @pool_name attribute.

derrickreimer committed:
Consolidate NodeJS adapter and its Config into one file

With the Adapters.Config macro gone, there's no longer a reason for
the NodeJS adapter's config struct to live in a separate module/file
— it has one consumer, is ~25 lines, and `@moduledoc false`. Move the
struct to a nested `Config` submodule inside `Inertia.SSR.Adapters.NodeJS`
(preserving the existing module path) and inline the `build/1` factory
into `init/1`. The `node_js/` directory disappears.

derrickreimer committed:
Drop NodeJS adapter Config struct in favour of a plain map

The struct's main wins (`@enforce_keys`, struct-tag pattern matching,
a typespec) aren't load-bearing for a value that's constructed in one
place and consumed in two — both right above each other in the same
module. Use a plain map instead and relax the behaviour's
`adapter_config` type to `term()`, which better reflects that
adapters are free to pick whatever shape suits them.

derrickreimer committed:
Loosen ssr_result error type to term()

`Inertia.SSR.Adapter.ssr_result` declared `{:error, String.t()}`, but
the underlying NodeJS.call/3 (and adapters in general) can return
non-string error values. The controller already defends against this
via an `is_binary(message)` check (added in df4e629), which dialyzer
flagged as unreachable given the over-narrow typespec.

Widen the error payload to `term()` so the type matches reality and
unblocks the defensive normalization in the controller.

derrickreimer committed:
Fold Bootstrap adapter resolution into Inertia.SSR

Bootstrap was @moduledoc false, had a single caller (Inertia.SSR.init/1),
and existed mostly as a holdover from earlier iterations. Inline its
adapter resolution as private helpers on Inertia.SSR and drop the file.

The three adapter-validation tests now call Inertia.SSR.init/1
directly — the real code path the supervisor takes — instead of the
intermediate Bootstrap.fetch_adapter/2. init/1 raises before any
:persistent_term.put or Supervisor.init runs, so failure-case tests
stay side-effect-free.

derrickreimer committed:
Rename Inertia.SSR.Adapters.NodeJS to Inertia.SSR.NodeJSAdapter

Switch from the Ecto-style `Adapters.*` namespace convention to the
Phoenix-style `XAdapter` suffix convention. With only one default
adapter shipped in-tree, the plural namespace was pure ceremony — the
`adapters/` directory held exactly one file. The flatter naming
(`Inertia.SSR.NodeJSAdapter` as a peer of `Inertia.SSR.Adapter`) is
also naming-symmetric with the singular behaviour module, and the
explicit `Adapter` suffix disambiguates from the `NodeJS` Hex package's
modules.

derrickreimer committed:
Preserve explicit ESM module filenames

The previous ESM path had two issues flagged by Codex review:

1. `module: "ssr.js"` (or `"ssr.mjs"`) with `esm: true` was rewritten
   to `"ssr.js.js"` / `"ssr.mjs.js"` — files that don't exist — because
   the `.js` extension was appended unconditionally.
2. `module: "ssr.mjs"` without an explicit `:esm` opt was passed to
   `NodeJS.call/3` with `esm: false`, overriding the upstream
   `:nodejs` package's auto-detection of ESM from a `.mjs` extension.

Fix both:

- Only append `.js` when the configured module has no extension; pass
  pre-extensioned modules through as-is.
- Default `:esm` to `nil` in the adapter state and only forward it to
  `NodeJS.call/3` when the user explicitly set it, letting the
  underlying package auto-detect from `.mjs`.

Add regression tests for both scenarios using the existing
`test/js/esm` fixture and a new `test/js/mjs/ssr.mjs`.

derrickreimer committed:
Merge branch '3.x' into feature/inertia-vitejs-integration

derrickreimer committed:
Sync SSR Adapter.page() type with current page object shape

The page object gained `flash` (now always present) plus optional keys
(`prependProps`, `deepMergeProps`, `matchPropsOn`, `onceProps`,
`scrollProps`, `sharedProps`, `preserveFragment`) that the `page()` type
never declared. Since a map type rejects unlisted keys, the map passed to
`SSR.call/1` broke the contract, which in turn made dialyzer mark the
downstream SSR functions unreachable.

derrickreimer committed:
Use data-inertia title marker in SSR test fixtures

The new ESM and MJS SSR fixtures emitted the legacy <title inertia>
marker instead of the v3 <title data-inertia> marker used elsewhere.
Aligning them means the SSR-rendered title now matches the controller's
title regex and exercises the same title-dedup path as the CJS fixture.
